### DISCUSSION TITLE
Encoder for numerical data instead of image_encoder

### DISCUSSION THREAD

**Discussion by mahsa-ebrahimian** [BindsNET/bindsnet#450] [Q&A] [answered]
Hi
I am totally new in using Spike neural network and I am developing model using bindsnet. The issue is that my dataset is numerical not image, so when creating class dataset, I see there are only image_encoders available. Could you please let me know how to encode my numerical data as spikes?

**Comment by RaisoLiu** [comment]
In example "eth_mnist", the step is:
image(value=0\~255) -> frequency(value=0\~(255/4)Hz) -> calculate spike time

So I think you can :
numerica(value=0\~max) -> frequency(value=0\~x Hz) -> calculate spike time

I'm new in this area. Don't believe me 100%.

**Comment by SimonInParis** [comment]
Hi.
To start with, you should consider rate-encoding. The Poisson encoder can
encode any scalar value into spike trains, not just images.
You can also have a rate based readout at the output layer.
Don't hesitate to look at the network examples, as they contain useful
mechanisms, even if they use images as inputs.
Simon

**Reply by mahsa-ebrahimian** [reply]
Thanks for your response, but I cant find any example of how to create a class of my own dataset, because the one I have just coded does not work well with the sample codes. I am following the supervised_mnist.py.

**Reply by mahsa-ebrahimian** [reply]
Actually I am stuck with loading my own dataset and encoding.
Here is the link to my code :  SNN model on SNN_model_on_MovieLens_dataset.ipynb

[https://github.com/mahsa-ebrahimian/Shilling_Attack_Detection](url)

**Reply by Hananel-Hazan** [reply]
I see that you almost there,  please change:
```
for (i, datum) in enumerate(dataloader):
    print(i)
```
To:

```
for (i, datum) in enumerate(range(len(dataloader))):
    print(i)
```

For more details on how to use DataLoader please cheak [here ](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html) and [here](https://stackoverflow.com/questions/62271194/pytorch-dataloader-from-csv-of-file-paths-and-labels)

Note, you need to normalize your dataset before feeding it to the network.

**Reply by mahsa-ebrahimian** [reply]
Thank you so much, it fixed the problem.

**Reply by mahsa-ebrahimian** [reply]
> I see that you almost there, please change:
> 
> ```
> for (i, datum) in enumerate(dataloader):
>     print(i)
> ```
> 
> To:
> 
> ```
> for (i, datum) in enumerate(range(len(dataloader))):
>     print(i)
> ```
> 
> For more details on how to use DataLoader please cheak [here ](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html) and [here](https://stackoverflow.com/questions/62271194/pytorch-dataloader-from-csv-of-file-paths-and-labels)
> 
> Note, you need to normalize your dataset before feeding it to the network.

Could you just let me know which example file or source you suggest for creating a SNN model for my classification problem on numerical dataset. I have been trying to use the supervised_mnist.py but its too complicated for my dataset.

**Reply by hockman1** [reply]
Same here. I would give 1million likes if someone could share a simple minimum viable working example on datasets such as sklearn iris dataset classification. It seems like the only problem is the data format conversion, otherwise, everything else is simple from there...

For example, there is this:

# Load MNIST data.
dataset = MNIST(
    PoissonEncoder(time=time, dt=dt),
    None,
    root=os.path.join("..", "..", "data", "MNIST"),
    download=True,
    transform=transforms.Compose(
        [transforms.ToTensor(), transforms.Lambda(lambda x: x * intensity)]
    ),
)

Based on what I understand, we need the training.pt and testing.pt. Does this mean we have to convert our own data files to this format in order for this Dataset Class to work?

**Reply by Hananel-Hazan** [reply]
Either convert it to this format or manually fetch each of your data point (training/testing set) and use one of the encoders to convert numeric value to a spike train. 

In any case, do a preprocess on your dataset and normalize it to a integer value between 0 to sum number.

**Reply by mahsa-ebrahimian** [reply]
> Either convert it to this format or manually fetch each of your data point (training/testing set) and use one of the encoders to convert numeric value to a spike train.
> 
> In any case, do a preprocess on your dataset and normalize it to a integer value between 0 to sum number.

I have tried manually fetching my data points and encode them using Poisson encoder, but I face the error:

RuntimeError: Index put requires the source and destination dtypes match, got Float for the destination and Double for the source.

Here is my code:

https://github.com/mahsa-ebrahimian/Shilling_Attack_Detection/blob/main/Supervised_MNIST_on_MovieLens.ipynb


@Hananel-Hazan   @SimonInParis

**Reply by Hananel-Hazan** [reply]
That is because your image type is double, try to cast your data to float, using something like:
`image = torch.tensor(image, dtype=torch.float)`

**Reply by mahsa-ebrahimian** [reply]
> That is because your image type is double, try to cast your data to float, using something like:
> `image = torch.tensor(image, dtype=torch.float)`

Thank you so much.
There is just one concern that when I apply encoding.poisson on my data it returns all the values of image equal to  zero.
I chose the time =100, do you have any idea why is it happening?
@Hananel-Hazan

**Reply by Hananel-Hazan** [reply]
Without knowing what is the input, I can only guess that the input values are too small. Try to normalize the data between 0 to the maximum running iteration of your network. Note that the data given to the Poisson encoder must be non-negative float.

**Reply by mahsa-ebrahimian** [reply]
> Without knowing what is the input, I can only guess that the input values are too small. Try to normalize the data between 0 to the maximum running iteration of your network. Note that the data given to the Poisson encoder must be non-negative float.

I tried increasing the time parameter and it seems to become fine. Does this parameter indicate the Poisson time intervals? How should it be chosen?

**Reply by Hananel-Hazan** [reply]
The Poisson function creates a spike train that you can feed to the network. The time parameter need to be the same as the time that the network spends on each input. 
There is no exact method how to choose this number. It depends on the input intensity, network activity and many other parameters. The best way to find the right parameter is by trial and error.

**Comment by mahsa-ebrahimian** [comment, marked as answer]
Thanks for all the responses I could **encode my data on Movie-Lens 100K dataset** using below steps:

1. Reading my numerical data from CSV using, model inputs are places in x and outputs in y:

```
X= pd.read_csv(os.path.join(data_dir, 'mydata_x.data'), '\t', names=names,
                           engine='python')

y= pd.read_csv(os.path.join(data_dir, 'mydata_y.data'), '\t', names=names,
                           engine='python')
```

2. Converting my x and y dataframe into tensors using:

```
x_tensor = torch.tensor(X.values, ,dtype=torch.float)
y_tensor = torch.tensor(y.values, ,dtype=torch.float)
```


3. **Encoding my numerical data using**:


```
from bindsnet         import encoding

for i, (image,label) in enumerate(zip(x_tensor ,y_tensor)):
    
    encoded_image=encoding.poisson(datum=image, time=time, dt=dt)
```

**Reply by mahsa-ebrahimian** [reply]
@hockman1  I hope my solution works for you too.
